keps/sig-instrumentation/20190605-metrics-validation-and-verification.md

```diff
@@ -0,0 +1,180 @@
+
+### Stable metric detection strategy 
+
+Static analysis will be done in two steps. First find stable metrics, second read and validate their fields. 
+
+In first one step we will want to distinguish stable metric without relying on their definition structure, allowing freedom of declaration for non-stable metrics. This will be done by finding occurrences of metric options object initialization (`CounterOpts` for `Counter`). If this initialization sets `StabilityLevel` then it will validate and read it. If value could not be extracted then script should fail. Metrics which set `kubemetrics.STABLE` will be passed to second step.
+
+Second step will be extracting information from stable metric definitions. This step will be expecting exact structure of AST subtree of new metric call matching format proposed above. If metric options object was initialized outside of new metric call or call structure deviates from expected format then static analysis should fail. This restriction will ensure that no field was read incorrectly or stable metric missed.
+
+
+### Format of stable metrics list file
+
+This file should include information about metrics needed to detect a breaking change. Those fields should include `name`, `namespace`, `subsystem` as they are used to build metric fully-qualified name, metric type and metric labels. Metric labels should include both vector labels and const labels. 
+
+Metric information stored in file should be in easly readable and reviewable format. For that pretty printed json will be used. Example:
+```json
+[
+  {
+    "name": "some_metric",
+    "subsystem": "some_subsystem",
+    "namespace": "some_namespace",
+    "labels": [
+      "some-label"
+    ],
+    "type": "Counter"
+  }
+]
+```
+
+Json keys and lists will be sorted alphabetically to ensure that there is only one correct output.
+
+### Failure Modes
+
+In cases where static analysis could not be done correctly (invalid stable metric format, etc.) script should fail (non zero response code) and write to STDERR information on problem. Example:
+```
+pkg/controller/volume/persistentvolume/scheduler_bind_cache_metrics.go:42 stable metric CounterOpts should be invoked from newCounter
+pkg/controller/volume/persistentvolume/scheduler_bind_cache_metrics.go:44 stable metric CounterOpts should be invoked from newCounter
+exit status 1
+```
+ 
+ ### Performance
+ 
+Proposed static analysis takes around 4 seconds on one CPU core. This time is comparable to conformance tests check ~2 sec and should be acceptable.
+
+ ## Graduation Criteria
+
+- [ ] Stable metric validation is enabled as pre-submit check
+
+ ## Implementation History
+
+- [ ] Implement static analysis script analysis all metrics types
+ 
+ ## Test Plan
+ 
+- [ ] Static analysis is covered by unit tests using common patterns of metric definition
+ 
+ ## Unresolved questions
+ 
+ ### Should changes in histogram buckets/summary objectives be a breaking change?
```

Comment:
Label value changes should be ingestion compatible.

Reply:
Discussion in other thread.